### DISCUSSION TITLE
CSW load issue in Terriajs8

### DISCUSSION THREAD

**Discussion by HeneyKellie** [TerriaJS/terriajs#5970] [Q&A] [answered]
We (@irishmarineinstitute @conbon27) are looking at implementing TerriaMap but are having issues bringing in CSW services. 

We have attempted to bring in multiple CSWs (from geonetwork sources) and all have returned the same error:
![image](https://user-images.githubusercontent.com/64596217/140541267-5dd0d7c7-8f3e-414e-94a6-15de193a217c.png)

The only CSWs that we have been able to successfully bring in are: https://ecat.ga.gov.au/geonetwork/srv/eng/csw?SERVICE=CSW&VERSION=2.0.2&REQUEST=GetCapabilities and https://marlin.csiro.au/geonetwork/srv/eng/csw?&request=GetCapabilities&service=CSW&version=2.0.2, otherwise we get the above error.

Specifically, the CSW services we are interested in bringing into TerriaMap are
1) https://data.marine.ie/geonetwork/srv/eng/csw?&request=GetCapabilities&service=CSW&version=2.0.2
2) https://isde.ie/geonetwork/srv/eng/csw?SERVICE=CSW&VERSION=2.0.2&REQUEST=GetCapabilities

We use these services in other situations without any issues. 
Is there something we are missing with our services, or is there a change we need to make somewhere else in our code to make these services work?

Thank you

**Comment by nf-s** [comment]
Hi @HeneyKellie 

Sorry this is a bug which was fixed a while ago, but the PR hasn't been merged - https://github.com/TerriaJS/terriajs/pull/5328
I tested both of your links and they now work - I will send you a test link when it has been merged.

Thanks!
Nick

**Comment by HeneyKellie** [comment]
Thanks @nf-s!

We have been able to add the CSWs for isde.ie and data.marine.ie into terriamap so that the metadata populates the data catalog, however we are experiencing a couple of different issues which result in the data not rendering on the map. 

1. 
![image](https://user-images.githubusercontent.com/64596217/144089236-dfed96e2-28ca-4c60-981b-ea7aba5583a8.png)

2. https services are loading into the map without symbology, or with symbology that is wildly distorted - but seems to maintain the underlying data.
![image](https://user-images.githubusercontent.com/64596217/144089515-1f01989b-b571-4492-9e2d-dcc7dc9969c9.png)
(This should not be a pink blob)

3. 
![image](https://user-images.githubusercontent.com/64596217/144089951-1ca76247-e6ae-4515-9212-9aaa76e3c174.png)
We have tried all sorts of adjustments and settings with the csw urls and the map service urls but the Tile Error persists

I have tried to follow some of the documentation here: https://docs.terria.io/guide/connecting-to-data/catalog-type-details/csw-group/ around setting up the csw groups, but I haven't had any success with addressing these above errors. 

Would this be down to our csw services themselves, or some settings in the init/config files that I am missing?
Any guidance you could provide would be greatly appreciated!
Thank you!
-Kellie

**Comment by nf-s** [comment, marked as answer]
Hi @HeneyKellie 

When CSW records are turned into WMS layers, Terria will look at the URI `name` attribute in the WMS URI

### Example record from [isde.ie](https://isde.ie/geonetwork/srv/eng/csw?service=CSW&version=2.0.2&request=GetRecordById&outputSchema=http%3A%2F%2Fwww.opengis.net%2Fcat%2Fcsw%2F2.0.2&ElementSetName=full&id=3a41c2fc-d529-4107-89cd-9dcbd913c62f)

This record has a correct WMS url, but the WMS name is `WMS`, therefore Terria will try to use the WMS lyaer `WMS`

<img width="645" alt="image" src="https://user-images.githubusercontent.com/6187649/145142036-74b6f107-5733-48f8-9a00-51774af7f9fc.png">

### Example record from [data.marine.ie](https://data.marine.ie/geonetwork/srv/eng/csw?service=CSW&version=2.0.2&request=GetRecordById&outputSchema=http%3A%2F%2Fwww.opengis.net%2Fcat%2Fcsw%2F2.0.2&ElementSetName=full&id=ie.marine.data%3Adataset.4408)

This record doesn't have a correct WMS url, and the WMS name is `Web Map Service`

<img width="653" alt="image" src="https://user-images.githubusercontent.com/6187649/145141695-de1c608c-3717-4194-808c-0358648a12e6.png">x

## Working example

### Configuration

```json
{
  "type": "csw-group",
  "name": "Marine Energy Context Layers",
  "description": "This group contains marine energy layers and other layers that can be used for decisions relating to marine energy devices.",
  "url": "http://oa-gis.csiro.au/geonetwork/srv/eng/csw",
  "domainSpecification": {
    "domainPropertyName": "awavea",
    "hierarchySeparator": " | ",
    "queryPropertyName": "subject"
  }
}
```

### Links

- http://oa-gis.csiro.au/geonetwork/srv/eng/csw?service=CSW&version=2.0.2&request=GetCapabilities
- Share link to [Marine Energy Context Layers in NationalMap](https://nationalmap.gov.au/#start=%7B%22version%22%3A%228.0.0%22%2C%22initSources%22%3A%5B%7B%22stratum%22%3A%22user%22%2C%22models%22%3A%7B%22Root+Group%2FNational+Data+Sets%2FEnergy%22%3A%7B%22isOpen%22%3Atrue%2C%22knownContainerUniqueIds%22%3A%5B%22Root+Group%2FNational+Data+Sets%22%5D%2C%22type%22%3A%22group%22%7D%2C%227Y8Bph%22%3A%7B%22isOpen%22%3Atrue%2C%22knownContainerUniqueIds%22%3A%5B%22Root+Group%2FNational+Data+Sets%2FEnergy%22%5D%2C%22type%22%3A%22group%22%7D%2C%22ZLc5P4%22%3A%7B%22isOpen%22%3Atrue%2C%22knownContainerUniqueIds%22%3A%5B%227Y8Bph%22%5D%2C%22type%22%3A%22group%22%7D%2C%22q5XBkW%22%3A%7B%22isOpen%22%3Atrue%2C%22knownContainerUniqueIds%22%3A%5B%22ZLc5P4%22%5D%2C%22type%22%3A%22group%22%7D%2C%22c2f37440%22%3A%7B%22isOpen%22%3Atrue%2C%22knownContainerUniqueIds%22%3A%5B%22q5XBkW%22%5D%2C%22type%22%3A%22csw-group%22%7D%2C%22rwwMKg%22%3A%7B%22isOpen%22%3Atrue%2C%22knownContainerUniqueIds%22%3A%5B%22Root+Group%2FNational+Data+Sets%2FEnergy%22%5D%2C%22type%22%3A%22group%22%7D%2C%22KrauUHsH%22%3A%7B%22isOpen%22%3Afalse%2C%22knownContainerUniqueIds%22%3A%5B%22Root+Group%2FNational+Data+Sets%22%5D%2C%22type%22%3A%22group%22%7D%2C%22vBwsllcv%22%3A%7B%22isOpen%22%3Afalse%2C%22knownContainerUniqueIds%22%3A%5B%22KrauUHsH%22%5D%2C%22type%22%3A%22wms-group%22%7D%2C%22idAbI6uY%22%3A%7B%22isOpen%22%3Atrue%2C%22knownContainerUniqueIds%22%3A%5B%22KrauUHsH%22%5D%2C%22type%22%3A%22group%22%7D%2C%22VhUa825H%22%3A%7B%22isOpen%22%3Atrue%2C%22knownContainerUniqueIds%22%3A%5B%22KrauUHsH%22%5D%2C%22type%22%3A%22group%22%7D%2C%22%2F%22%3A%7B%22type%22%3A%22group%22%7D%2C%22Root+Group%2FNational+Data+Sets%22%3A%7B%22knownContainerUniqueIds%22%3A%5B%22%2F%22%5D%2C%22type%22%3A%22group%22%7D%7D%2C%22workbench%22%3A%5B%5D%2C%22timeline%22%3A%5B%5D%2C%22initialCamera%22%3A%7B%22west%22%3A105.08706096560452%2C%22south%22%3A-39.53366118489588%2C%22east%22%3A161.91293952384936%2C%22north%22%3A-9.036618004587874%2C%22position%22%3A%7B%22x%22%3A-6733654.97437972%2C%22y%22%3A7095791.782500638%2C%22z%22%3A-4863110.017932153%7D%2C%22direction%22%3A%7B%22x%22%3A0.6155631636719867%2C%22y%22%3A-0.6486682277326119%2C%22z%22%3A0.44756175200788473%7D%2C%22up%22%3A%7B%22x%22%3A-0.30808117990014067%2C%22y%22%3A0.32464982435187667%2C%22z%22%3A0.8942530280293342%7D%7D%2C%22homeCamera%22%3A%7B%22west%22%3A109%2C%22south%22%3A-45%2C%22east%22%3A158%2C%22north%22%3A-8%7D%2C%22baseMaps%22%3A%7B%22defaultBaseMapId%22%3A%22basemap-positron%22%7D%2C%22viewerMode%22%3A%223d%22%2C%22currentTime%22%3A%7B%22dayNumber%22%3A2459556%2C%22secondsOfDay%22%3A53020.472%7D%2C%22showSplitter%22%3Afalse%2C%22splitPosition%22%3A0.5%2C%22previewedItemId%22%3A%22c2f37440%22%2C%22stories%22%3A%5B%5D%7D%5D%7D)

### Example record from [oa-gis.csiro.au](http://oa-gis.csiro.au/geonetwork/srv/eng/csw?service=CSW&version=2.0.2&request=GetRecordById&outputSchema=http%3A%2F%2Fwww.opengis.net%2Fcat%2Fcsw%2F2.0.2&ElementSetName=full&id=c4bd11c5c880d8c72af33c01ac1b4b5fe0fae33b)

Note: highlighted `name` attribute, this is used as WMS layer

<img width="650" alt="image" src="https://user-images.githubusercontent.com/6187649/145141258-2c62148f-9bb4-4348-8df6-3b3285dd11af.png">

Let me know if you need more info!

Cheers
